### PR TITLE
api: Adding Lock helpers for leader election

### PR DESCRIPTION
**Pull request by armon**
This adds a new `Lock` to the client API. This can be used to implement simple leader election or mutual exclusion via Consul. Modeled on: https://consul.io/docs/guides/leader-election.html

### CODE REVIEW

File: api/lock.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+	l.l.Lock()
+	defer l.l.Unlock()
+
+	// Ensure the lock is actually held
+	if !l.isHeld {
+		return ErrLockNotHeld
+	}
+
+	// Set that we no longwer own the lock
+	l.isHeld = false
```
Comment (ryanuber) [inline api/lock.go]: I think we need to set this just before returning. Otherwise if we fail to release the lock, this gets set to false anyways.

**Comment by ryanuber** [comment]
@armon I left 1 comment, but otherwise everything LGTM! Super awesome.

File: api/lock.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+	// Hold the lock as we try to release
+	l.l.Lock()
+	defer l.l.Unlock()
+
+	// Ensure the lock is actually held
+	if !l.isHeld {
+		return ErrLockNotHeld
+	}
+
+	// Set that we no longwer own the lock
```
Comment (blalor) [inline api/lock.go]: Typo. `s/longwer/longer/`. :-)

File: api/lock.go (hunk 1 above)
Comment (armon) [inline api/lock.go]: The only thing that can fail is the KV.Release() but we have to assume the lock is no longer held once that call is started.

**Pull request merged**